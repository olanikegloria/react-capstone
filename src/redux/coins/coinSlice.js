import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

const initialState = {
  coins: [],
  loading: true,
  error: false,
  searchQuery: '',
  singleCoin: null, // Change to null initially
  coinStat: '',
  singleCoinChanged: false, // New field to track changes
};

const options = {
  method: 'GET',
  url: 'https://coinranking1.p.rapidapi.com/coins',
  params: {
    referenceCurrencyUuid: 'yhjMzLPhuIDl',
    timePeriod: '24h',
    'tiers[0]': '1',
    orderBy: 'marketCap',
    orderDirection: 'desc',
    limit: '20',
    offset: '0',
  },
  headers: {
    'X-RapidAPI-Key': 'a018ef236dmshdee85e2adbcdef9p1e359djsn1b163cdcb0dc',
    'X-RapidAPI-Host': 'coinranking1.p.rapidapi.com',
  },
};

export const fetchCoin = createAsyncThunk('coins/fetchCoin',
  async (uuid, thunkAPI) => {
    try {
      const response = await axios.request(`https://coinranking1.p.rapidapi.com/coin/${uuid}`, options);
      console.log(response);
      return response.data;
    } catch (error) {
      return thunkAPI.rejectWithValue('Something went wrong');
    }
  });

export const getCoins = createAsyncThunk('coins/getCoins',
  async (thunkAPI) => {
    try {
      const response = await axios.request(options);
      return response.data;
    } catch (error) {
      return thunkAPI.rejectWithValue('Something went wrong');
    }
  });

const coinsSlice = createSlice({
  name: 'coins',
  initialState,
  reducers: {
    setSearch: (state, action) => {
      state.searchQuery = action.payload;
    },
    setSingleCoin: (state, action) => {
      state.singleCoinChanged = state.singleCoin !== action.payload;
      state.singleCoin = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(getCoins.pending, (state) => {
        state.loading = true;
      })
      .addCase(getCoins.fulfilled, (state, action) => {
        state.loading = false;
        state.coins = action.payload.data.coins;
      })
      .addCase(getCoins.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      })
      .addCase(fetchCoin.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchCoin.fulfilled, (state, action) => {
        state.loading = false;
        state.singleCoin = action.payload.data.coin;
      })
      .addCase(fetchCoin.rejected, (state, action) => {
        state.loading = false;
        state.error = action.payload;
      });
  },
});

export const { setSearch, setSingleCoin } = coinsSlice.actions;
export default coinsSlice.reducer;
