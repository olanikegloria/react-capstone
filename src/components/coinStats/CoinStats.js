import React, { useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { fetchCoin, setSingleCoin } from '../../redux/coins/coinSlice';
import '../coin.css';
import Coinstat from './coinstat';

function CoinStats() {
  const { coinId } = useParams();
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchCoin(coinId));
  }, [dispatch, coinId]);

  useEffect(() => {
    // Dispatch an action to update the singleCoin value
    dispatch(setSingleCoin(/* new singleCoin value */));
  }, [dispatch]);

  const { singleCoin } = useSelector((state) => state.coins);
  return (
    <div>
      <Coinstat
        singleCoin={singleCoin}
      />
    </div>
  );
}

export default CoinStats;
