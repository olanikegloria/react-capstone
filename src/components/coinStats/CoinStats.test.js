import React from 'react';
import { act, create } from 'react-test-renderer'; // Import the act and create functions
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import CoinStats from './CoinStats';
import { fetchCoin, setSingleCoin } from '../../redux/coins/coinSlice';

const mockStore = configureStore([]);
const store = mockStore({
  coins: {
    singleCoin: {
      uuid: 'Qwsogvtv82FCd',
      name: 'Bitcoin',
      symbol: 'BTC',
      // Other mock properties...
    },
  },
});

jest.mock('../../redux/coins/coinSlice', () => ({
  fetchCoin: jest.fn(),
  setSingleCoin: jest.fn(),
}));

// ... (other imports and mock setups)

describe('CoinStats component', () => {
  it('renders correctly with mock data', () => {
    const expectedUuid = 'Qwsogvtv82FCd'; // Update with the actual uuid being passed

    act(() => {
      const component = (
        <Provider store={store}>
          <CoinStats />
        </Provider>
      );

      // Render the component using the create function
      const tree = create(component);

      // Assert that the component renders correctly
      expect(tree.toJSON()).toMatchSnapshot();

      // Assert that the fetchCoin action was dispatched with the correct argument
      expect(fetchCoin).toHaveBeenCalledWith(expectedUuid);

      // Assert that the setSingleCoin action was dispatched with the correct argument
      expect(setSingleCoin).toHaveBeenCalledWith({
        uuid: 'Qwsogvtv82FCd',
        name: 'Bitcoin',
        symbol: 'BTC',
        // Other mock properties...
      });
    });
  });
});
