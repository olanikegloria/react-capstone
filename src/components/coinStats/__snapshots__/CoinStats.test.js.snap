// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinStats component renders correctly with mock data 1`] = `null`;
